src/jit/decomposelongs.cpp

```diff
@@ -628,8 +628,20 @@ GenTree* DecomposeLongs::DecomposeCast(LIR::Use& use)
         }
         else
         {
-            if (varTypeIsUnsigned(srcType))
+            //
+            // If the cast is not used or it is used by a GT_MUL then we'll treat it as unsigned
+            // because doing so generates simpler IR. If a GT_MUL reaches here then it must be
+            // the GTF_MUL_64RSLT variant that uses only the low parts of its long operands, see
+            // DecomposeMul for details.
+            //
+
+            if (varTypeIsUnsigned(srcType) || use.IsDummyUse() || (use.User()->OperGet() == GT_MUL))
             {
+                if (!use.IsDummyUse() && (use.User()->OperGet() == GT_MUL))
```

Comment:
Fair enough. Hopefully the C++ compiler is smart enough to erase the if statement in a release build. :)